### PR TITLE
nrf91: improvements for modem and support for modem sockets

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary

- include/nuttx/wireless/lte/lte.h: increase bb_product length
- include/nuttx/wireless/lte/lte.h: add fw_version field to lte_version_t
- arch/nrf91: fix modem initialization
- arch/nrf91: convert modem AT interface to serial device
- arch/nrf91: fixes for modem OS integration
- arch/nrf91: add support for modem sockets
- boards/nrf9160-dk: update modem configuration

## Testing
nrf9160-dk

```
NuttShell (NSH) NuttX-10.4.0
nsh> cu -l /dev/modem            # we have to set the device to full functionality, here we use direct AT commands
AT+CFUN=1
OK
AT+CFUN?
+CFUN: 1
OK
~.
nsh>
nsh> wget http://www.example.com
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        div {
            margin: 0 auto;
            width: auto;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This domain is for use in illustrative examples in documents. You may use this
    domain in literature without prior coordination or asking for permission.</p>
    <p><a href="https://www.iana.org/domains/example">More information...</a></p>
</div>
</body>
</html>
nsh>

```

### CODE REVIEW

File: arch/arm/src/nrf91/nrf91_modem.h [hunk 1]
```diff
@@ -32,12 +34,17 @@
...
+#define HEAP_OVERHEAD          (400)
+
 /* Shared memory configuration */
 
 #define NRF91_SHMEM_START_ADDR (CONFIG_RAM_START)
 #define NRF91_SHMEM_CTRL_BASE  (NRF91_SHMEM_START_ADDR)
 #define NRF91_SHMEM_TX_BASE    (NRF91_SHMEM_CTRL_BASE+NRF91_SHMEM_CTRL_SIZE)
-#define NRF91_SHMEM_TX_SIZE    (CONFIG_NRF91_MODEM_SHMEM_TX_SIZE)
+#define NRF91_SHMEM_TX_SIZE    (CONFIG_NRF91_MODEM_SHMEM_TX_SIZE\
+                                - HEAP_OVERHEAD)
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf91/nrf91_modem.h:47]: merge into one line?

File: include/nuttx/wireless/lte/lte.h [hunk 2]
```diff
@@ -195,6 +195,10 @@
 
 #define LTE_VER_NP_PACKAGE_LEN (32)
 
+/* Length of character string for firmware version */
+
+#define LTE_VER_FIRMWARE_LEN     (32)
```
Comment (xiaoxiang781216) [inline include/nuttx/wireless/lte/lte.h:200]: remove the extra spaces

File: arch/arm/src/nrf91/nrf91_modem_at.c [hunk 3]
```diff
@@ -112,140 +157,215 @@ static void nrf91_modem_at_notify_handler(const char *notif)
...
+/****************************************************************************
+ * Name: nrf91_modem_at_rxavailable
+ ****************************************************************************/
 
-  nxmutex_unlock(&dev->lock);
-  return ret;
+static bool nrf91_modem_at_rxavailable(struct uart_dev_s *dev)
+{
+  struct nrf91_modem_at_s *priv = (struct nrf91_modem_at_s *)dev->priv;
+  return (priv->notif || priv->resp);
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf91/nrf91_modem_at.c:277]: remove ()

File: arch/arm/src/nrf91/nrf91_modem.h (hunk 1 above)
Comment (raiden00pl) [reply arch/arm/src/nrf91/nrf91_modem.h]: done

File: arch/arm/src/nrf91/nrf91_modem_at.c (hunk 3 above)
Comment (raiden00pl) [reply arch/arm/src/nrf91/nrf91_modem_at.c]: done

File: include/nuttx/wireless/lte/lte.h (hunk 2 above)
Comment (raiden00pl) [reply include/nuttx/wireless/lte/lte.h]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**